ldms/src/ldmsd/ldmsd_config.c

```diff
@@ -771,6 +771,214 @@ int __process_config_file(const char *path, int *lno, int trust,
+		goto parse;
+	}
+
+	tmp[cnt] = '\0';
+
+	/* Get rid of leading spaces */
+	while (isspace(*tmp)) {
+		tmp++;
+		cnt--;
+	}
+
+	if (!cnt) {
+		/* empty buffer */
+		goto parse;
+	}
+
+	if (tmp[cnt-1] == '\\') {
+		if (cnt == 1)
+			goto parse;
+	}
+
+	if (cnt + off > line_sz) {
+		char *nline = realloc(line, ((cnt + off)/line_sz + 1) * line_sz);
+		if (!nline) {
+			rc = errno;
+			ldmsd_log(LDMSD_LERROR, "Out of memory\n");
+			goto cleanup;
+		}
+		line = nline;
+		line_sz = ((cnt + off)/line_sz + 1) * line_sz;
+	}
+	off += snprintf(&line[off], line_sz, "%s", tmp);
+
+	/* attempt to merge multiple lines together */
+	if (off > 0 && line[off-1] == '\\') {
+		line[off-1] = ' ';
+		goto next_line;
+	}
+
+parse:
+	if (!off)
+		goto next_line;
+
+	if (ldmsd_is_initialized()) {
+		if ((0 == strncmp(line, "prdcr_add", 9)) ||
+				(0 == strncmp(line, "prdcr_start", 11))) {
+			if (strstr(line, "interval")) {
+				ldmsd_log(LDMSD_LWARNING,
+						"'interval' is being deprecated. "
+						"Please use 'reconnect' with 'prdcr_add' or 'prdcr_start*' "
+						"in the future.\n");
+			}
+		}
+	}
+
+	req_array = ldmsd_parse_config_str(line, lineno, xprt.max_msg, ldmsd_log);
+	if (!req_array) {
+		rc = errno;
+		ldmsd_log(LDMSD_LERROR, "Process config string error in line\n %s\n "
+				". %s\n", line, STRERROR(rc));
+		goto cleanup;
+	}
+
+	request = __aggregate_records(req_array);
+	if (!request) {
+		rc = errno;
+		goto cleanup;
+	}
+	ldmsd_req_array_free(req_array);
+	req_array = NULL;
+
+	if (!ldmsd_is_initialized()) {
+		/* Process only the priority commands, e.g., cmd-line options */
+		if (!is_req_id_priority(ntohl(request->req_id)))
+			goto next_req;
+	} else {
+		/* Process non-priority commands, e.g., cfgobj config commands */
+		if (is_req_id_priority(ntohl(request->req_id)))
+			goto next_req;
+	}
+
+	/*
+	 * Make sure that LDMSD will create large enough buffer to receive
+	 * the config data.
+	 */
+	if (xprt.max_msg < ntohl(request->rec_len))
+		xprt.max_msg = ntohl(request->rec_len);
+
+	rc = ldmsd_process_config_request(&xprt, request, req_filter, ctxt);
+	if (rc || xprt.rsp_err) {
+		if (!rc)
+			rc = xprt.rsp_err;
+		ldmsd_log(LDMSD_LERROR, "Configuration error at line %d (%s)\n",
+				lineno, cfg_str);
+		goto cleanup;
+	}
+next_req:
+	free(request);
+	request = NULL;
+	off = 0;
+	goto next_line;
+
+cleanup:
+	if (buff)
+		free(buff);
+	if (line)
+		free(line);
+	if (lno)
+		*lno = lineno;
+	ldmsd_req_array_free(req_array);
+	if (request)
+		free(request);
+	return rc;
+}
+
+char *__process_yaml_config_file(const char *path, const char *dname)
+{
+	FILE *fp;
+	char command[256];
+	char cstr[256];
+        char *cfg_str = malloc(512);
+	*cfg_str = '\0';
+	snprintf(command, sizeof(command), "ldmsd_yaml_parser --ldms_config %s --daemon_name %s 2>&1", path, dname);
+	fp = popen(command, "r");
+	if (!fp)
+		printf("Error in yaml_parser\n");
+	size_t char_cnt = 0;
+	int status;
+	while (fgets(cstr, sizeof(cstr), fp) != NULL) {
```

Comment:
The logic here looks broken in a few different ways to me. Once we cross over 512 bytes, it'll start realloc()ing, every single time, but only increasing the buffer size by 1 byte. But then we strcat something that is potentially alot larger than one byte larger.

But lets forget that. Fundamentally using fgets()/strlen()/strcat() are the wrong way to read a file into a buffer. This needs to be rewritten to use fread(). We can directly fread() into our buffer (no extra copies needed), and resize only when the buffer fills between fread()s.